parsons/google/google_bigquery.py

```diff
@@ -1388,6 +1388,93 @@ def extract(
+            destination_table: str
+                Name of destination table
+            if_dataset_not_exists: str
+                Action if dataset doesn't exist {'fail','create'}
+            if_table_exists: str
+                Action if table exists {'fail', 'overwrite'}
+
+        `Returns:`
+            None
+        """
+
+        from google.cloud import bigquery
+        from google.cloud.exceptions import NotFound
+        import sys
+
+        destination_table_id = (destination_project + "."
+                                + destination_dataset + "."
+                                + destination_table)
+        source_table_id = (source_project + "."
+                           + source_dataset + "."
+                           + source_table)
+
+        # check if destination dataset exists
+        try:
+            dataset_id = destination_project + "." + destination_dataset
+            self.client.get_dataset(dataset_id)  # Make an API request.
+            # if it exists: continue
+        except NotFound:
+            # if it doesn't exist: check the flag for creating it
+            if if_dataset_not_exists == 'create':  # create a new dataset in the destination
+                dataset = bigquery.Dataset(dataset_id)
+                dataset = self.client.create_dataset(dataset, timeout=30)
+            else:  # fail
+                print("BigQuery copy failed")
+                sys.exit("Dataset {0} does not exist and if_dataset_not_exists set to {1}"
+                         .format(destination_dataset, if_dataset_not_exists))
+
+        job_config = bigquery.CopyJobConfig()
+
+        # check if destination table exists
+        try:
+            # check if table exists
+            self.client.get_table(destination_table_id)
+            if if_table_exists == 'overwrite':  # if it exists
+                job_config = bigquery.CopyJobConfig()
+                job_config.write_disposition = "WRITE_TRUNCATE"
+                try:
+                    job = self.client.copy_table(source_table_id, destination_table_id,
+                                                 location='US', job_config=job_config)
+                    result = job.result()
+                except Exception as e:
+                    print("Exception during copy between projects", e)
+            else:
+                print("BigQuery copy failed")
+                sys.exit("Table {0} exists and if_table_exists set to {1}"
+                         .format(destination_table, if_table_exists))
+
+        except NotFound:
+            # table doesn't exist
+            try:
+                job = self.client.copy_table(source_table_id, destination_table_id,
+                                             location='US', job_config=job_config)
+                result = job.result()
+                print(result)
+            except Exception as e:
+                print("Exception during copy between projects", e)
```

Comment:
While it makes sense to capture the NotFound exceptions and log the results without actually raising the exception, this exception here just captures any exception and saves the feedback. Generally speaking we want to raise any unexpected errors so we can notice and fix whatever the problem is sooner rather than later, so I would remove this specific try/except block and just call the code normally. If there's an exception, there's an exception - and we want to know about it!

Reply:
Generally speaking, in software development, "don't catch general exceptions" is a good rule of thumb. [This blog post](https://jerrynsh.com/stop-using-exceptions-like-this-in-python/) has some good details of why.